sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -271,32 +265,24 @@ static BigtableSegmentReaderImpl create(BigtableSession session, BigtableSource
               .setFilter(filter)
               .setRowsLimit(source.getMaxBufferElementCount())
               .build();
-      return new BigtableSegmentReaderImpl(
-          session, request, source.getTableId().get(), filter, source.getMaxBufferElementCount());
+
+      serviceCallMetric = populateReaderCallMetric(session, source.getTableId().get());
+
+      long maxSegmentByteSize =
+          (long)(Runtime.getRuntime().totalMemory()
+              * DEFAULT_BYTE_LIMIT_PERCENTAGE);
+
+      return new BigtableSegmentReaderImpl(session, request, maxSegmentByteSize);
     }
 
     @VisibleForTesting
-    BigtableSegmentReaderImpl(
-        BigtableSession session,
-        ReadRowsRequest request,
-        String tableId,
-        RowFilter filter,
-        int segmentSize) {
+    BigtableSegmentReaderImpl(BigtableSession session, ReadRowsRequest request, long maxSegmentByteSize) {
```

Comment:
I think you can move everything in this constructor, something like: 

```
static BigtableSegmentReaderImpl create(BigtableSession session, BigtableSource source) {
    long maxSegmentByteSize = ...
    return new BigtableSegmentReaderImpl(session, request, maxSegmentByteSize);
} 

@VisibleForTesting
BigtableSegmentReaderImpl(BigtableSession session, BigtableSource source, long maxSegmentByteSize) {
      RowSet set;
      if (source.getRanges().isEmpty()) {
        set = RowSet.newBuilder().addRowRanges(RowRange.getDefaultInstance()).build();
      } else {
      ....
}
```

Reply:
Igor wanted to avoid using source when we don't have to. I'm okay either way, @igorbernstein2 thoughts?

Reply:
yea, I want to minimize the surface of the source that the reader needs to know about